Get/set n-th bit of `BigUint` and `BigInt`

janmarthedal opened the pull request:
This PR implements `bit` and `set_bit` for `BigUint` and `BigInt`.

The method names have been chosen to match those of Ramp (https://docs.rs/ramp/0.5.9/ramp/int/struct.Int.html#method.bit).

For `BigInt` the implementation uses the number's two's complement representation when the number is negative. This matches what libraries like Ramp or languages like Python do.

Resolves #172

cuviper commented on src/bigint.rs:
```diff
@@ -3254,6 +3254,70 @@ impl BigInt {
+                } else {
+                    // This is the general case that basically corresponds to what `bitor_neg_pos`
+                    // (when setting bit) or `bitand_neg_pos` (when clearing bit) does, except there
+                    // is no need to explicitly iterate over the digits of the right-hand side
```
I have to admit, that bit-twiddling has fallen out of my memory by now. Your adaptation looks appropriate, but I can't help feeling that there must be a way to shorten the scan both before and after the target bit. That is, before only needs to figure out the carry state, and after might be able to stop on zero carry. I'm not sure.

Maybe this is a good candidate for some fuzzing? There are tests of this sort in `ci/big_quickcheck`.

janmarthedal replied:
You are absolutely correct, in most cases it will iterate over far more than necessary. When writing the code, I was in doubt of whether I should go for more efficient and complicated code, or sub-optimal code that was fairly straightforward and shorter. I went for the latter, but I sort of regret that now. I am currently rewriting the `set_bit` method, so stay tuned!

cuviper reviewed (commented):
Thanks! The two's complement stuff is a bit hairy, as expected, but I don't see any problem with it.

janmarthedal commented:
@cuviper I have now updated the PR with a more efficient implementation of `BigInt::set_bit`. Basically, I have made sure that only the relevant digits will be iterated over instead of all of them, as you also pointed out. In a special benchmark it saved about 70% in runtime.

I have also implemented a special case for `bit < trailing_zeros && value` which also saves some runtime compared to the case `bit == trailing_zeros && !value`, but only about 5-10%. The two cases can be merged together to make the `set_bit` a bit shorter.

mkaczanowski commented:
@cuviper it's been a while since the last review :) Is this PR ready? (asking, since I am waiting for this feature)

cuviper commented:
I added a couple tweaks to the edge cases of `u64` to `usize`, but otherwise it looks good!

bors r+

cuviper commented:
bors r+
